pkg/machine/compression/generic.go

```diff
@@ -0,0 +1,68 @@
+package compression
+
+import (
+	"io"
+	"io/fs"
+	"os"
+
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/sirupsen/logrus"
+)
+
+type genericDecompressor struct {
+	compressedFilePath     string
+	compressedFile         *os.File
+	decompressedFileReader io.ReadCloser
+	compressedFileInfo     os.FileInfo
+}
+
+func newGenericDecompressor(compressedFilePath string) (*genericDecompressor, error) {
+	d := &genericDecompressor{}
+	d.compressedFilePath = compressedFilePath
+	stat, err := os.Stat(d.compressedFilePath)
+	if err != nil {
+		return nil, err
+	}
+	d.compressedFileInfo = stat
+	return d, nil
+}
+
+func (d *genericDecompressor) compressedFileSize() int64 {
+	return d.compressedFileInfo.Size()
+}
+
+func (d *genericDecompressor) compressedFileMode() fs.FileMode {
+	return d.compressedFileInfo.Mode()
+}
+
+func (d *genericDecompressor) compressedFileReader() (io.ReadCloser, error) {
+	compressedFile, err := os.Open(d.compressedFilePath)
+	if err != nil {
+		return nil, err
+	}
+	d.compressedFile = compressedFile
+	return compressedFile, nil
+}
+
+func (d *genericDecompressor) decompress(w io.WriteSeeker, r io.Reader) error {
+	decompressedFileReader, _, err := compression.AutoDecompress(r)
+	if err != nil {
+		return err
+	}
+	d.decompressedFileReader = decompressedFileReader
+
+	_, err = io.Copy(w, decompressedFileReader)
+	return err
+}
+
+func (d *genericDecompressor) close() {
+	if err := d.compressedFile.Close(); err != nil {
+		logrus.Errorf("Unable to close compressed file: %q", err)
+	}
+
+	if d.decompressedFileReader != nil {
```

Comment:
I think this can be closed (using `defer`) already in `decompress`. Similarly in the other implementations.

That would also shorten the various subclasses of `genericDecompressor`.

Reply:
I should change the `decompress` method to return `(io.Closer, error)` but not all the decompressed streams implement the io.Closer interface (e.g. `*zstd.Decoder` doesn't).

Reply:
I mean, the most natural way to ensure “open” and “close” calls match is to do both in the same function, ideally
```go
foo := newSomething(…)
defer foo.Close()
```

AFAICS none of the decompressors _need_ the close to be done only in the separate `Close`.

Reply:
I am sorry but I didn't get your suggestion at first. That makes sense now. I have done the changes to `defer` the `close()` right in the `decompress` methods. That indeed simplifies the code. Thanks!